File: ml-service/graphsage/employee_baseline.py
"""
Employee Behavioral Baseline Engine (Phase 59)
Constructs personal historical behavioral baselines to detect "unusual for this employee" patterns point-in-time.
"""

import numpy as np
from datetime import datetime, timezone, timedelta
from typing import Dict, List, Any, Optional, Set
from collections import Counter, defaultdict

from .schema import HeteroEdge, HeteroNode


class EmployeeBehavioralBaselineEngine:
    """
    Computes per-employee baseline profiles point-in-time (strictly events < T).
    """

    def compute_baseline_and_deviation(
        self,
        employee_id: str,
        current_events: List[HeteroEdge],
        historical_edges: List[HeteroEdge],
        as_of: datetime
    ) -> Dict[str, Any]:
        """
        Calculates personal baseline profile and deviation scores.
        """
        # Filter historical edges strictly before evaluation window
        prior_edges = [
            e for e in historical_edges
            if (e.source_id == employee_id or e.target_id == employee_id)
            and e.timestamp < as_of
        ]

        if not prior_edges:
            # Cold-start employee fallback
            return {
                "has_historical_baseline": False,
                "prior_interaction_count": 0,
                "mean_daily_volume": 0.0,
                "normal_active_hours": [9, 10, 11, 12, 13, 14, 15, 16, 17],
                "trusted_devices": [],
                "trusted_ips": [],
                "volume_z_score": 0.0,
                "is_unusual_hour_for_employee": False,
                "is_new_device_for_employee": True,
                "personal_anomaly_delta": 0.15,
                "baseline_notes": "Zero historical telemetry for this employee; default role baseline applied."
            }

        # 1. Historical hours distribution
        hours_hist = [e.timestamp.hour for e in prior_edges]
        hour_counts = Counter(hours_hist)
        frequent_hours = [h for h, cnt in hour_counts.items() if (cnt / len(prior_edges)) >= 0.05]

        # 2. Historical daily volumes
        day_buckets = defaultdict(int)
        for e in prior_edges:
            day_key = e.timestamp.strftime("%Y-%m-%d")
            day_buckets[day_key] += 1

        daily_vols = list(day_buckets.values())
        mean_vol = float(np.mean(daily_vols)) if daily_vols else 1.0
        std_vol = float(np.std(daily_vols)) + 1e-5

        # 3. Trusted infrastructure
        trusted_devs = set()
        trusted_ips = set()
        for e in prior_edges:
            etype = e.type.value if hasattr(e.type, "value") else str(e.type)
            if "DEVICE" in etype:
                trusted_devs.add(e.target_id if e.source_id == employee_id else e.source_id)
            if "IP" in etype:
                trusted_ips.add(e.target_id if e.source_id == employee_id else e.source_id)

        # 4. Compare Current Activity against Historical Baseline
        curr_hours = [e.timestamp.hour for e in current_events]
        curr_unusual_hours = sum(1 for h in curr_hours if h not in frequent_hours)
        is_unusual_hour = (curr_unusual_hours / max(len(current_events), 1)) >= 0.50 if current_events else False

        curr_vol = float(len(current_events))
        vol_z = (curr_vol - mean_vol) / std_vol

        # Check new devices
        curr_devs = set()
        for e in current_events:
            etype = e.type.value if hasattr(e.type, "value") else str(e.type)
            if "DEVICE" in etype:
                curr_devs.add(e.target_id if e.source_id == employee_id else e.source_id)

        is_new_device = any(d not in trusted_devs for d in curr_devs) if curr_devs else False

        # Combined personal deviation metric
        anomaly_delta = 0.0
        if is_unusual_hour:
            anomaly_delta += 0.35
        if vol_z > 2.0:
            anomaly_delta += 0.30
        if is_new_device and len(trusted_devs) > 0:
            anomaly_delta += 0.25

        return {
            "has_historical_baseline": True,
            "prior_interaction_count": len(prior_edges),
            "mean_daily_volume": round(mean_vol, 2),
            "std_daily_volume": round(std_vol, 2),
            "normal_active_hours": sorted(frequent_hours),
            "trusted_devices": list(trusted_devs),
            "trusted_ips": list(trusted_ips),
            "volume_z_score": round(float(vol_z), 2),
            "is_unusual_hour_for_employee": bool(is_unusual_hour),
            "is_new_device_for_employee": bool(is_new_device),
            "personal_anomaly_delta": round(float(min(1.0, anomaly_delta)), 4),
            "baseline_notes": f"Baseline constructed from {len(prior_edges)} events across {len(daily_vols)} active days."
        }

File: backend/internal/graph/graphsage/employee_baseline.go
package graphsage

import (
	"fmt"
	"math"
	"time"
)

// EmployeeBaselineProfile captures personal historical habits.
type EmployeeBaselineProfile struct {
	HasHistoricalBaseline bool      `json:"has_historical_baseline"`
	PriorEventCount       int       `json:"prior_event_count"`
	MeanDailyVolume       float64   `json:"mean_daily_volume"`
	FrequentHours         []int     `json:"frequent_hours"`
	TrustedDevices        []string  `json:"trusted_devices"`
	TrustedIPs            []string  `json:"trusted_ips"`
	VolumeZScore          float64   `json:"volume_z_score"`
	IsUnusualHour         bool      `json:"is_unusual_hour"`
	IsNewDevice           bool      `json:"is_new_device"`
	PersonalAnomalyDelta  float64   `json:"personal_anomaly_delta"`
	Notes                 string    `json:"notes"`
}

// EmployeeBaselineEngine computes point-in-time personal baseline profiles in Go.
type EmployeeBaselineEngine struct{}

// ComputeBaselineAndDeviation evaluates current activity against historical profile.
func (e *EmployeeBaselineEngine) ComputeBaselineAndDeviation(
	employeeID string,
	currentEvents []*HeteroEdge,
	historicalEdges []*HeteroEdge,
	asOf time.Time,
) *EmployeeBaselineProfile {
	var prior []*HeteroEdge
	for _, edge := range historicalEdges {
		if (edge.SourceID == employeeID || edge.TargetID == employeeID) && edge.Timestamp.Before(asOf) {
			prior = append(prior, edge)
		}
	}

	if len(prior) == 0 {
		return &EmployeeBaselineProfile{
			HasHistoricalBaseline: false,
			FrequentHours:         []int{9, 10, 11, 12, 13, 14, 15, 16, 17},
			PersonalAnomalyDelta:  0.10,
			Notes:                 "Cold-start employee; default standard baseline applied.",
		}
	}

	hourCounts := make(map[int]int)
	daysMap := make(map[string]int)
	trustedDevs := make(map[string]bool)
	trustedIPs := make(map[string]bool)

	for _, edge := range prior {
		h := edge.Timestamp.Hour()
		hourCounts[h]++
		dKey := edge.Timestamp.Format("2006-01-02")
		daysMap[dKey]++

		if edge.Type == EdgeEmployeeUsesDevice || edge.Type == EdgeUsesDevice {
			target := edge.TargetID
			if target == employeeID {
				target = edge.SourceID
			}
			trustedDevs[target] = true
		}
		if edge.Type == EdgeEmployeeUsesIP || edge.Type == EdgeUsesIP {
			target := edge.TargetID
			if target == employeeID {
				target = edge.SourceID
			}
			trustedIPs[target] = true
		}
	}

	var frequentHours []int
	for h, cnt := range hourCounts {
		if float64(cnt)/float64(len(prior)) >= 0.05 {
			frequentHours = append(frequentHours, h)
		}
	}

	meanVol := float64(len(prior)) / math.Max(float64(len(daysMap)), 1.0)
	currVol := float64(len(currentEvents))
	volZ := (currVol - meanVol) / math.Max(meanVol*0.5, 1.0)

	freqHourMap := make(map[int]bool)
	for _, h := range frequentHours {
		freqHourMap[h] = true
	}

	unusualHourCount := 0
	isNewDev := false
	for _, edge := range currentEvents {
		if !freqHourMap[edge.Timestamp.Hour()] {
			unusualHourCount++
		}
		if edge.Type == EdgeEmployeeUsesDevice {
			target := edge.TargetID
			if target == employeeID {
				target = edge.SourceID
			}
			if len(trustedDevs) > 0 && !trustedDevs[target] {
				isNewDev = true
			}
		}
	}

	isUnusual := false
	if len(currentEvents) > 0 && float64(unusualHourCount)/float64(len(currentEvents)) >= 0.50 {
		isUnusual = true
	}

	anomalyDelta := 0.0
	if isUnusual {
		anomalyDelta += 0.35
	}
	if volZ > 2.0 {
		anomalyDelta += 0.30
	}
	if isNewDev {
		anomalyDelta += 0.25
	}
	anomalyDelta = math.Min(1.0, anomalyDelta)

	var devList []string
	for d := range trustedDevs {
		devList = append(devList, d)
	}
	var ipList []string
	for ip := range trustedIPs {
		ipList = append(ipList, ip)
	}

	return &EmployeeBaselineProfile{
		HasHistoricalBaseline: true,
		PriorEventCount:       len(prior),
		MeanDailyVolume:       math.Round(meanVol*100) / 100,
		FrequentHours:         frequentHours,
		TrustedDevices:        devList,
		TrustedIPs:            ipList,
		VolumeZScore:          math.Round(volZ*100) / 100,
		IsUnusualHour:         isUnusual,
		IsNewDevice:           isNewDev,
		PersonalAnomalyDelta:  math.Round(anomalyDelta*100) / 100,
		Notes:                 fmt.Sprintf("Profile computed from %d historical events across %d days.", len(prior), len(daysMap)),
	}
}
